ext/drreg/drreg.dox

```diff
@@ -42,7 +42,13 @@ instrumentation pass scheme.  The results are most efficient when used in
 the \p drmgr instrumentation insertion phase.  Use outside of the insertion
 phase, such as during \p drmgr's instrumentation-to-instrumentation
 transformation phase, or use outside of \p drmgr entirely, is supported,
-but duplicate spills and restores may not be optimized away.
+but duplicate spills and restores may not be optimized away.  When using
+drreg to reserve registers in multiple phases (like app2app or
+instru2instru in addition to the insertion phase), it is important to call
+drreg_set_bb_properties() with #DRREG_HANDLE_MULTI_PHASE_SLOT_RESERVATIONS
+to enable logic that avoids conflicts with spill slots used by previous
+phases.  This may have some performance cost, so we do not turn it on by
+default.
```

Comment:
Another issue is that the slots used in app2app are now permanently unavailable for insertion.  Many clients tune their register slot usage precisely (TLS slots are limited commodities on Windows).   Please add a sentence or two here making that fact explicit.

We then need to change the drx_init code to ask for additional *summed* slots, or if we don't want to pay for that for drx in general if nobody is using drx_expand_* then we need to document that clients have to do this in their drreg init.  How many slots does drx_expand_* use?  Should we be worried about too many slots on Windows?
See drx_init which looks like it's only asking for drx_insert_counter_update today and is *not* summing:
```
     * We set do_not_sum_slots to true so that we only ask for *more* slots
     * if the client doesn't ask for any.
```

Reply:
> How many slots does drx_expand_* use?

To find this, I used `drreg_max_slots_used` which keeps track of max spill slots used so far. Before encountering a scatter/gather, it seems to be just 2 (probably from drcachesim/other). The first time a scatter/gather is expanded, it goes up to 4. And subsequently, it goes up to 6 (4 reserved from scatter/gather expansion in app2app, and new 2 from drcachesim/other).
This matches what I see in code (one in `expand_avx512_scatter_gather_update_mask` and 3 in `drx_expand_scatter_gather` itself)

`drx_expand_scatter_gather` already asks for more spill slots
https://github.com/DynamoRIO/dynamorio/blob/9319132573ab39bfe292f6850c69d16b7c71c365/ext/drx/drx.c#L2322
But I'm confused by the total 6 it requests and the comment there. As `do_not_sum_slots` is true, I guess it's specifying the total?

Reply:
Hmm, that `!expand_scatter_gather_drreg_initialized` code looks suspect:
+ It's racy
+ It calls drreg_init potentially quite late, while drreg_init docs say it must be called before any other drreg routines, and as we were discussing it really should be limited to process init

Reply:
>  if we don't want to pay for that for drx in general if nobody is using drx_expand_* then we need to document that clients have to do this in their drreg init.

I'm leaning towards paying for it in drx for all cases. It doesn't sound reliable to ask clients to be aware of the instrs they are tracing beforehand.

I see that we have some heuristics in core DR that enable expensive saving/restoring of zmm/opmask regs when we see:
- that the client was compiled with avx flags: static check, could work for us, but may not cover all cases
- some avx instr in a basic block: dynamic check. not known at process init

How big a concern is having an extra 4 spill slots all the time?

Reply:
> How big a concern is having an extra 4 spill slots all the time?

See https://github.com/DynamoRIO/dynamorio/blob/master/core/lib/dr_tools.h#L1972

See #1163.  Large Windows apps can use up the available slots all by themselves, so DR taking any number can cause apps to fail to run.

We added -alt_teb_tls but it looks like it's used last: maybe we should change that and have it used first to have fewer app conflicts: though it's a little hacky, and on Windows we have late injection so all the libs have already grabbed their slots, so maybe it doesn't matter as much.

I definitely agree that it is much simpler to completely separate the app2app slots have drx request the max at init time.  The only issue is Windows TLS space.

It's not the end of the world to fall back on DR slots: they're just slower beyond the first 3 w/ an extra indirection load.

One option would be to request 4 on UNIX and on Windows fall back on DR slots -- except the app2app phase is going to take all the direct slots and other libs/client slots that might be more perf-critical might end up being the DR slots?

Or just take 4 on Windows too and see if anyone complains.  The biggest use of DR on Windows is drcov: if drx were used by drcov I would be concerned, but it's not.

Reply:
> One option would be to request 4 on UNIX and on Windows fall back on DR slots 

This sounds reasonable, as Windows is known to be much more sensitive to available slot count.

> -- except the app2app phase is going to take all the direct slots and other libs/client slots that might be more perf-critical might end up being the DR slots?

The app2app phase is going to take all direct slots only for the scatter/gather instr bbs. Note that scatter/gather instrs have their own separate bb. So the perf overhead should be limited to only these basic blocks.

Reply:
Done.

Reply:
I didn't understand the 6 spill slot issue: does it end up using 6 for scatter-gather (that seems like an awful lot)?

Reply:
`drx_expand_scatter_gather` itself seems to be using 4; that's what I see on using `drreg_max_slots_used` to get the value at runtime, and comparing before/after `drx_expand_scatter_gather`.

2 others are being used elsewhere (should be drcachesim insertion phase, like `drx_insert_counter_update`). As insertion and app2app phase cannot share spill slots, that's a total of 6.

Reply:
Discussed offline that we'll need to observe perf impact and fine-tune slot reservations if needed. There are unknowns in its impact on Windows and large clients like DrMemory. Good to submit for now though.